lib/IPCMsgQSend.cpp

```diff
@@ -0,0 +1,169 @@
+	}
+
+	mq_unlink(this->opts.server_name.c_str());
+	std::cout << "Unlinked the message queue name." << std::endl;
+}
+
+void IPCMsgQSend::cleanup()
+{
+	if (this->mqd > 0)
+	{
+		mq_close(this->mqd);
+		std::cout << "Closed the message queue." << std::endl;
+	}
+
+	mq_unlink(this->opts.server_name.c_str());
+	std::cout << "Unlinked the message queue name." << std::endl;
+	IPC::cleanup();
+}
+
+void IPCMsgQSend::init()
+{
+	// remove old queue name (if any)
+	mq_unlink(this->opts.server_name.c_str());
+	errno = 0; // clear errno
+	this->mqd = mq_open(this->opts.server_name.c_str(), O_CREAT | O_EXCL | O_WRONLY | O_NONBLOCK, 0660, &(this->attr));
+
+	if (this->mqd == -1)
+	{
+		cleanup();
+		throw std::runtime_error("ERROR: " + this->opts.server_name + ": " + strerror(errno));
+	}
+	else
+		std::cout << "/dev/mqueue" << this->opts.server_name << " is opened." << std::endl;
+
+	this->open_file();
+	this->ipc_info.file_size = this->get_file_size();
+	if(this->ipc_info.file_size == 0)
+	{
+		this->cleanup();
+		throw std::runtime_error("ERROR: File size = 0.");
+	}
+	this->buffer = new char[this->attr.mq_msgsize];
+	this->timer.update_all();
+}
+
+void IPCMsgQSend::print_members() const
+{
+	IPC::print_members();
+	std::cout << "mqd: " << this->mqd << "\n"
+			  << "attr.mq_flags: " << this->attr.mq_flags << "\n"
+			  << "attr.mq_msgsize: " << this->attr.mq_msgsize << "\n"
+			  << "attr.mq_maxmsg: " << this->attr.mq_maxmsg << "\n"
+			  << "attr.mq_curmsgs: " << this->attr.mq_curmsgs << "\n"
+			  << "mq_priority: " << this->mq_priority << std::endl;
+}
+
+void IPCMsgQSend::send()
+{
+	long mq_send_return_value{0};
+
+	std::cout << "Sending..." << std::endl;
+	while (!this->fs.eof())
+	{
+		this->fs.read(this->buffer, this->attr.mq_msgsize);
+		if (this->fs.bad()) // check read/writing error on i/o operation
+			throw std::runtime_error("ERROR: istream::read().");
+
+		this->ipc_info.read_bytes = this->fs.gcount();
+		if (this->ipc_info.read_bytes > 0)
+		{
+			errno = 0; // clear errno
+			mq_send_return_value = mq_send(this->mqd, this->buffer, this->ipc_info.read_bytes, this->mq_priority);
+			if (mq_send_return_value == 0)
+			{
+				this->ipc_info.total_sent_bytes += this->ipc_info.read_bytes;
+				++(this->ipc_info.number_of_msg);
+				this->timer.update_all();
+			}
+			else if (mq_send_return_value == -1 && errno == EAGAIN)
+			{
+				while (errno == EAGAIN && this->timer.get_duration() < this->timeout)
+				{
+					print_wait_msg("The message queue is full. Waiting for client to empty the queue");
+					this->timer.update_end();
+					errno = 0; // clear errno
+					mq_send_return_value = mq_send(this->mqd, this->buffer, this->ipc_info.read_bytes,
+												   this->mq_priority);
+					if (mq_send_return_value == 0)
+					{
+						this->ipc_info.total_sent_bytes += this->ipc_info.read_bytes;
+						++(this->ipc_info.number_of_msg);
+						this->timer.update_all();
+					}
+					sleep(1);
```

Comment:
I understand the way you are doing this. But I think it's really noneffective to sleep 1s each time the queue is empty. Usually, the receiver should take a message more quickly than 1 second.

Reply:
True, I just want to put it there to demonstrate how the method works :D